transports/quic/src/hole_punching.rs

```diff
@@ -0,0 +1,126 @@
+use std::{
+    collections::HashMap,
+    net::SocketAddr,
+    pin::Pin,
+    sync::{Arc, Mutex},
+    task::{Context, Poll},
+    time::Duration,
+};
+
+use futures::{
+    channel::oneshot,
+    future::{Fuse, FusedFuture},
+    prelude::*,
+};
+use futures_timer::Delay;
+use libp2p_identity::PeerId;
+use rand::{distributions, Rng};
+
+use crate::{
+    endpoint::{self, ToEndpoint},
+    Connecting, Connection, Error,
+};
+
+pub(crate) type HolePunchMap =
```

Comment:
Instead of a single synchronized map on `(SocketAddr, PeerId)` you could as well have two maps, one nested in the other:

1. Inside `GenTransport` an unsynchronized `HashMap<SocketAddr, Arc<Mutex<HashMap<PeerId, oneshot::Sender<(PeerId, Connection)>>>>`.
2. Inside `MaybeHolePunchedConnection` and `Listener` the inner `Arc<Mutex<HashMap<PeerId, oneshot::Sender<(PeerId, Connection)>>>`.

`MaybeHolePunchedConnection` and `Listener` don't need to know about all ongoing hole punches, but only the ones concerning their `SocketAddr`.

This would remove the need for synchronization across all hole punches and it would remove the need for `MaybeHolePunchedConnection` to know its `addr`.

@arpankapoor what do you think?

Reply:
`SocketAddr` in the hashmap key is the remote address and not the listener address, so `Listener` can not store the inner `Arc<Mutex<HashMap<PeerId, oneshot::Sender<(PeerId, Connection)>>>` as you suggest.

Reply:
Right. My bad. Thanks for explaining.